v5 Release

angularsen opened the pull request:
Fixes #180 

Merging the v5 release branch.

It is still in alpha, but it is functional, nugets are published and there are not many planned breaking changes left.
By merging, all efforts moving forward are targeting v5 and this reduces friction:

- No more merge conflicts trying to forward port all changes to v5, instead cherry pick new units and fixes to v4 until v5 is fully stable.
- Contributors are having trouble building v4 locally due to `net40`, `net47` and Windows Runtime Component targets.


## 💥 Breaking changes
Default number format should be CultureInfo.CurrentCulture, not CurrentUICulture (#795)
Use CurrentCulture rather than CurrentUICulture (#986)
Return `QuantityValue` in `IQuantity` properties instead of `double` (#1074)
Return `decimal` in properties of `Power`, `BitRate` and `Information` quantities (#1074)
Fix singular name VolumeFlow.MillionUsGallonsPerDay

## 🔥 Removed 
Remove targets: net40, net47, Windows Runtime Component.
Remove `Undefined` enum value for all unit enum types
Remove QuantityType enum
Remove IQuantity.Units and .UnitNames
Remove IEquatable<T> and equality operators/methods 
Remove GlobalConfiguration
Remove obsolete and deprecated code.
Remove Molarity ctor and operator overloads
Remove MinValue, MaxValue per quantity due to ambiguity
Remove string format specifiers: "v", "s"
json: Remove UnitsNetJsonConverter

## ✨ New
QuantityValue: Implement IEquality, IComparable, IFormattable
QuantityValue: 16 bytes instead of 40 bytes (#1084)
Add `[DataContract]` annotations (#972)

## ♻️ Improvements
Upgrade CodeGen, tests and sample apps to net6.0.

## 📝 JSON unit definition schema changes
Rename `BaseType` to `ValueType`, for values "double" and "decimal".
Rename `XmlDoc` to `XmlDocSummary`.

## TODO
Add back `IEquatable<T>`, but implement as strict equality with tuple of quantity name + unit + value.
https://github.com/angularsen/UnitsNet/issues/1017#issuecomment-1028401955

## Postponed for later
#1067

angularsen commented:
@tmilnthorp @lipchev The V5 release branch is rolling.
If there is anything you are eager to include/remove, then please give a heads up here. I hope to make this release in not too long.

I'll be going through #180 to see what seems like low-hanging fruit and try to get those, but don't wait for me if you want to do some of them.

One thing I'm looking to remove is UnitSystem, BaseUnits and BaseDimension as per #630 and #651, until the design is more complete, easier to use and maybe extensible. I'm interested to know if you are using this today so that removing it would cause problems for you.

lipchev commented:
I think we can include the GetHashCode and IComparable tests/fixes in #838 (even if we removed IEquatable).

lipchev commented:
As far as the UnitSystems - last I checked the implementation was ok for review. Or at least as a starting point (e.g. for v5). I didn't remove the BaseDimensions, but that could be done in a consequent step (with little change the current implementation). The only major thing missing was editing the remaining json files (adding default SI value).
I don't have a particular need for it right now, but I think there could be potential uses for it- even beyond just the standard unit system.

lipchev commented:
Also, if we would be breaking things- please consider also #983

ebfortin commented:
I don't know if it's here that we should discuss breaks in v5 but here I go. Yesterday I upgraded to v5 alpha 3 to test it on one of my project and I got an error for the == operator. Going back to 4.x resolved the issue.

angularsen commented:
@ebfortin Thanks, could you please report back with the exact error and what you would expect instead?
There are many breaking changes in v5, but if you found something that seems incorrect then it would be good to know.

ebfortin commented:
> @ebfortin Thanks, could you please report back with the exact error and what you would expect instead? There are many breaking changes in v5, but if you found something that seems incorrect then it would be good to know.

Severity	Code	Description	
Error	CS0019	Operator '==' cannot be applied to operands of type 'ElectricPotential' and 'ElectricPotential'

angularsen commented:
@ebfortin Right, it's in the breaking changes list.
> Remove IEquatable and equality operators/methods

The reasoning is that comparing equality of two quantities should use the `Equals()` methods, where you can specify the accepted error. Internally, the quantity may be represented by double or decimal, so it is prone to subtle rounding errors.

ebfortin commented:
Hmmm ok I see. At first I wrote that we should keep == and IEquatable since we have == for doubles and when you use it you know there are rounding errors possible. But I just realized that in my use case I shouldn't use == so the very reason you removed it. So thanks for that. I'll change my code and update to 5.

Télécharger Outlook pour Android<https://aka.ms/AAb9ysg>

________________________________
De : Andreas Gullberg Larsen ***@***.***>
Envoyé : vendredi 26 novembre 2021 20 h 35
À : angularsen/UnitsNet
Cc : Étienne Fortin; Mention
Objet : Re: [angularsen/UnitsNet] v5 Release (PR #982)


@ebfortin<https://github.com/ebfortin> Right, it's in the breaking changes list.

Remove IEquatable and equality operators/methods

The reasoning is that comparing equality of two quantities should use the Equals() methods, where you can specify the accepted error. Internally, the quantity may be represented by double or decimal, so it is prone to subtle rounding errors.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/angularsen/UnitsNet/pull/982#issuecomment-980482892>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABNYDWOEYHMW5D2LXKCMMWTUOAYXFANCNFSM5HFGGE7A>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

tmilnthorp commented:
I didn't see this sorry.

> One thing I'm looking to remove is UnitSystem, BaseUnits and BaseDimension

I think ```UnitSystem``` can go, but ```BaseDimensions``` and ```BaseUnits``` are useful and I'd like to see them stay. ```BaseDimensions``` and ```BaseUnits``` apply to all units. They help you know things like ```Length``` * ```Length```= ```Area``` as well as the conversions needed for different units in that calculation (e.g. 2m * 2in = 0.1016 m^2).

angularsen commented:
> I think `UnitSystem` can go, but `BaseDimensions` and `BaseUnits` are useful and I'd like to see them stay

Thanks, that is useful to know.
I am open to keep it, but I feel it would help me to learn what it can be used for in practice. I have not used them myself and am not sure what some good use cases are.

I find tests on BaseDimensions and BaseUnits isolated, but I can't seem to find any tests or examples where they are used together in a meaningful way. Some notes on the below points would help describe why they are useful and what future work may look like. Then we could document it in the wiki for future reference.

- What are some practical use cases for BaseDimensions/BaseUnits as it stands today? Or with future work?
- We can multiply base dimensions `Length^1 * Length^1 = Length^2`, but how do we arrive at `Area`?
- Isn't there ambiguity in that multiple quantities share the same base dimensions? F.ex. ApparentEnergy, Energy, RotationalStiffness and Torque seems to have L=2, M=1, T=-2.
- Some quantities don't seem to map to base dimensions, such as Angle or VitaminA. Is it fine to just leave them out?



If we keep them, how about renaming them to `SiBaseDimensions` and `SiBaseUnits`?

tmilnthorp commented:
> > I think `UnitSystem` can go, but `BaseDimensions` and `BaseUnits` are useful and I'd like to see them stay
> 
> Thanks, that is useful to know. I am open to keep it, but I feel it would help me to learn what it can be used for in practice. I have not used them myself and am not sure what some good use cases are.
> 
> I find tests on BaseDimensions and BaseUnits isolated, but I can't seem to find any tests or examples where they are used together in a meaningful way. Some notes on the below points would help describe why they are useful and what future work may look like. Then we could document it in the wiki for future reference.
> 
> * What are some practical use cases for BaseDimensions/BaseUnits as it stands today? Or with future work?
> * We can multiply base dimensions `Length^1 * Length^1 = Length^2`, but how do we arrive at `Area`?
> * Isn't there ambiguity in that multiple quantities share the same base dimensions? F.ex. ApparentEnergy, Energy, RotationalStiffness and Torque seems to have L=2, M=1, T=-2.
> * Some quantities don't seem to map to base dimensions, such as Angle or VitaminA. Is it fine to just leave them out?
> 
> If we keep them, how about renaming them to `SiBaseDimensions` and `SiBaseUnits`?

It was for future work, sorry I dropped off for a bit. Hoping to come back here 😃 

To answer your questions:
> What are some practical use cases for BaseDimensions/BaseUnits as it stands today? Or with future work?

I'd like to use BaseDimensions/BaseUnits in Multiply/Divide methods (and later in code like #984). We can actually generate these statically. For example, you know that Length * Length = Area. This uses BaseDimensions. For BaseUnits, Let's take Speed for example. We currently have a custom method: ```public static Speed operator /(Length length, Duration duration)```. This could be generated dynamically. There's also the issue that this will only return ```Speed.FromMetersPerSecond```. That isn't too useful if you passed a Length of ```Feet``` and want like-units. Base units would let you dynamically construct 1000ft / 1s = 1000fps (```Speed.FeetPerSecond```)

> We can multiply base dimensions Length^1 * Length^1 = Length^2, but how do we arrive at Area?

You find the Quantities where ```Info.BaseDimensions.Length``` == 2.

> Isn't there ambiguity in that multiple quantities share the same base dimensions? F.ex. ApparentEnergy, Energy, RotationalStiffness and Torque seems to have L=2, M=1, T=-2.

Yes. That is where things get tricky. We need to figure something out there, as you can't overload based on return type. We could just vary the method names.

> Some quantities don't seem to map to base dimensions, such as Angle or VitaminA. Is it fine to just leave them out?

```VitaminA``` is an odd one, I will say. It is in fact a ```Mass``` for a particular substance, so it does have a base unit. ```Angle``` is correctly identified as ```Dimensionless```.

angularsen commented:
Added a TODO:
Add back `IEquatable<T>`, but implement as strict equality with tuple of quantity name + unit + value.
https://github.com/angularsen/UnitsNet/issues/1017#issuecomment-1028401955

angularsen commented:
@tmilnthorp What do you think about this?
> If we keep them, how about renaming them to SiBaseDimensions and SiBaseUnits?

Primarily want to distinguish from Length.BaseUnit and "BaseUnit" in JSON.
I also think it helps to be explicit that these are related SI unit system.

tmilnthorp commented:
> @tmilnthorp What do you think about this?
> 
> > If we keep them, how about renaming them to SiBaseDimensions and SiBaseUnits?
> 
> Primarily want to distinguish from Length.BaseUnit and "BaseUnit" in JSON. I also think it helps to be explicit that these are related SI unit system.

These are not related to SI. **All** units have dimensionality and base units

- Inches is a base quantity (non-SI obviously)
  - Dimensions
    - Length: 1
  - Base unit
    - Inch
- Square inches is a derived quantity
  - Dimensions
    - Length: 2
  - Base unit
    - Inch
- Miles per hour is a derived quantity (length / time)
  - Dimensions (length / time)
    - Length: 1
    - Time: -1
  - Base units
    - Length: miles
    - Time: hour

For 60mph you can see the miles unit goes in the numerator (length dimension of +1) and hours go in the denominator (time dimension of -1).

See https://en.m.wikipedia.org/wiki/International_System_of_Quantities#Derived_quantities  and https://en.wikipedia.org/wiki/Dimensional_analysis

tmilnthorp commented:
I wonder if we should name ```Length.BaseUnit``` something like ```ReferenceUnit```. That's really what we have: a reference unit (```Meter``` for ```Length```) to do conversions from ```Inch``` -> ```Meter``` -> ```Mile``` (for example).

angularsen commented:
> I wonder if we should name `Length.BaseUnit` something like `ReferenceUnit`

Yes, I initially tried ConversionBaseUnit, but it's just cumbersome and long so I reverted it.
ReferenceUnit is a decent alternative.

Regarding SI, you are probably right. It just seemed to me that there exists multiple unit systems beyond SI, and I was probing whether we should consider supporting anything other than SI? 
I can't really find any units systems that are widely used today. Do you know of any?

Unit systems: 
- [CGS: centimeters-grams-second.](https://en.wikipedia.org/wiki/Centimetre%E2%80%93gram%E2%80%93second_system_of_units) Introduced in 1832, predecessor of SI.
- [MKS: meter-kilogram-second.](https://en.wikipedia.org/wiki/MKS_system_of_units) Introduced in the 1940s, predecessor of SI, successor to CGS.
- [SI](https://en.wikipedia.org/wiki/International_System_of_Units#:~:text=The%20SI%20comprises%20a%20coherent,candela%20(cd%2C%20luminous%20intensity)): Introduced in the 1960s.
- [FPS: foot–pound–second.](https://en.wikipedia.org/wiki/Foot%E2%80%93pound%E2%80%93second_system#:~:text=The%20foot%E2%80%93pound%E2%80%93second%20system,and%20the%20second%20for%20time.) and its 3 variants: British Graviational, English Engineering, Absolute English. Replaced by SI?


If there are no other unit systems we would like to support, then I guess the naming `BaseUnits` would suffice and be unambiguous.
